### PR TITLE
feat: [CDS-77359]: add adviser obtainments for failure strategies

### PR DESCRIPTION
**Pull request by shalini-agr**
## Describe your changes

**Changes:** Adding advisers for failure strategies for v1 for HTTP Step. Putting failure strategies in parent info in the Custom stage, step group plan creator.

**Testing:** Tested failure strategies at the HTTP step, custom stage, and step group level. Types of failure strategies tested: abort, ignore, success, fail, and manual intervention.

TODO: Retry, Retry Step Group FS,Stage and Pipeline rollback, proceed with default values.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52355)

### CODE REVIEW

**Comment by shalini-agr** [comment]
trigger uts

**Comment by shalini-agr** [comment]
trigger uts

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52355)* status: 4 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/group/GroupPlanCreatorV1.java` line 92 at r2](https://reviewable.io/reviews/harness/harness-core/52355#-NdUSgfN9FlNK1yH1UZ-:-NdUSgfN9FlNK1yH1UZ0:b-46ic5m) ([raw file](https://github.com/harness/harness-core/blob/ef4d6b5fac7848a6e0cdcd22e549cae0c98213f1/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/group/GroupPlanCreatorV1.java#L92)):*
> ```Java
>   }
> 
>   Dependency getDependencyForChildren(PlanCreationContext ctx, YamlField config) {
> ```

we should re-discuss this approach with @hextechpal as one con with this approach this we are duplicating same info for all steps, which will lead to memory increased and grpc data transfer for cross steps.

**Review by hextechpal — APPROVED** [review]

**Review by brijeshdhakar — DISMISSED** [review]
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52355)* status: 6 file(s) to review. 1 change requests.

File: clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java [hunk 1]
```diff
@@ -195,21 +198,47 @@ private AdviserObtainment getManualInterventionAdviserObtainment(KryoSerializer
...
-    Optional<Object> stepGroupFailureStrategiesOptional = getDeserializedObjectFromDependency(
-        dependency, kryoSerializer, PlanCreatorConstants.STEP_GROUP_FAILURE_STRATEGIES, true);
-    OnConfigV1 stepGroupFailureStrategies = null;
-    if (stepGroupFailureStrategiesOptional.isPresent()) {
-      stepGroupFailureStrategies = (OnConfigV1) stepGroupFailureStrategiesOptional.get();
+    return stageFailureStrategies;
+  }
+
+  OnConfigV1 getStepGroupFailureStrategies(KryoSerializer kryoSerializer, Dependency dependency) {
+    Optional<Object> stageFailureStrategiesOptional = getDeserializedObjectFromParentInfo(
```
Comment (utkarshchoubey-harness) [inline clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java:235]: nit- name it stepGroupFailureStrategiesOptional

**Review by utkarshchoubey-harness — APPROVED** [review]

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52355)* status: 1 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java` line 212 at r4](https://reviewable.io/reviews/harness/harness-core/52355#-Nd_GaGdEyykOgBLDU2A:-Nd_GaGdEyykOgBLDU2B:b-884y0) ([raw file](https://github.com/harness/harness-core/blob/72d28a6dc9e535c38782e40d432e55055a033a21/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java#L212)):*
> ```Java
>     if (dependency != null && dependency.getParentInfo() != null
>         && dependency.getParentInfo().getDataMap().containsKey(key)
>         && dependency.getParentInfo().getDataMap().get(key) != null) {
> ```

can this if condition be a private method? difficult to read in current method and gives more good code read if separated



_Code quote:_
```Java
    if (dependency != null && dependency.getParentInfo() != null
        && dependency.getParentInfo().getDataMap().containsKey(key)
        && dependency.getParentInfo().getDataMap().get(key) != null) {
```
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java` line 219 at r4](https://reviewable.io/reviews/harness/harness-core/52355#-Nd_GhX4An1ooBS3Am8w:-Nd_GhX4An1ooBS3Am8x:by945v0) ([raw file](https://github.com/harness/harness-core/blob/72d28a6dc9e535c38782e40d432e55055a033a21/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java#L219)):*
> ```Java
>         }
>         return Optional.of(kryoSerializer.asObject(bytes.toByteArray()));
>       }
> ```

this piece of code is repeated, can be extracted out

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52355)* status: 2 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java` line 212 at r4](https://reviewable.io/reviews/harness/harness-core/52355#-Nd_GaGdEyykOgBLDU2A:-Nd_LpYB3j0Z18Uy-5bh:b-do95x3) ([raw file](https://github.com/harness/harness-core/blob/72d28a6dc9e535c38782e40d432e55055a033a21/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java#L212)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

can this if condition be a private method? difficult to read in current method and gives more good code read if separated
</blockquote></details>

simplified this condition by removing the redundant ones
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java` line 219 at r4](https://reviewable.io/reviews/harness/harness-core/52355#-Nd_GhX4An1ooBS3Am8w:-Nd_Ltjl7BE6cnWpT_Lg:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/72d28a6dc9e535c38782e40d432e55055a033a21/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/PlanCreatorUtilsV1.java#L219)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

this piece of code is repeated, can be extracted out
</blockquote></details>

Done.
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/group/GroupPlanCreatorV1.java` line 92 at r2](https://reviewable.io/reviews/harness/harness-core/52355#-NdUSgfN9FlNK1yH1UZ-:-Nd_LvZA1vvAqDUnpNVc:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/ef4d6b5fac7848a6e0cdcd22e549cae0c98213f1/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/group/GroupPlanCreatorV1.java#L92)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

we should re-discuss this approach with @hextechpal as one con with this approach this we are duplicating same info for all steps, which will lead to memory increased and grpc data transfer for cross steps.
</blockquote></details>

done

**Review by archit-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52355)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by mohitg0795 — DISMISSED** [review]

**Comment by shalini-agr** [comment]
trigger uts

**Comment by shalini-agr** [comment]
trigger review

**Comment by shalini-agr** [comment]
trigger ut5, ut6, ut8, ut9

**Comment by shalini-agr** [comment]
trigger ut7

**Comment by shalini-agr** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Comment by shalini-agr** [comment]
trigger uts

**Comment by shalini-agr** [comment]
trigger uts

**Review by mohitg0795 — APPROVED** [review]

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52355)* status: 1 file(s) to review.

**Pull request merged**